forcedotcom/salesforcedx-vscode
# [WIP] build: use a bundler (esbuild) along with vsce to package vsix files

### What does this PR do?
Uses esbuild to bundle (and treeshake) JS files before packaging them into vsix with vsce. This would lead to 
- significant reduction in file sizes leading to faster downloading and potentially faster loading of extensions for users
- breaking dependence on npm v6 for vsce package which is [currently blocking us from upgrading to npm v7/v8 (node 16)](https://github.com/forcedotcom/salesforcedx-vscode/pull/4067) or [pnpm](https://github.com/forcedotcom/salesforcedx-vscode/pull/4042) 
- Continues https://github.com/forcedotcom/salesforcedx-vscode/pull/3748

### What issues does this PR fix or reference?
@W-6980161@

### Functionality Before
- vsce package was used to produce vsix files

### Functionality After
- vsce package is used along with esbuild to produce vsix files

### Notes
- Currently `npm prune --production` is run before vsce package 
  - This removes all dev dependencies 
    - Hence current packaging is a destructive operation requiring install and compile to continue develop
  - vsce package then gets a list of files to be packaged from the package manager (auto-detected)
  - Without npm prune, vsce packaging errors complaining about the dev deps e.g. with `npm ERR! extraneous: typescript@3.8.3` 
    - Even using a tool such as [duniul/clean-modules: 🧹 Clean up/prune unnecessary files and reduce the size of your node_modules directory.](https://github.com/duniul/clean-modules) in place of npm prune doesn't work
    
#### Future considerations 👉 
- QA
   - Even when bundling isn't working correctly (missing to identify files to bundle due to dynamic imports e.g.) the vsix files build and install without any errors
      - Have to watch for extension activation errors in VSCode dev console, ext host log etc
      - Without automated e2e tests to verify correct activation and e2e functionality of the extensions, bundling could risk the quality of the releases
- Debugging / user error reporting 
  - Stack traces become harder to read and less informative with bundled packages
